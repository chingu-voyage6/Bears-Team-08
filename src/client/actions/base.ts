import { Action } from "./action";

export type NavToExplorer = {
  type: "NAV_TO_EXPLORER";
};

export const navToExplorer = (): Action => ({
  type: "NAV_TO_EXPLORER"
});
