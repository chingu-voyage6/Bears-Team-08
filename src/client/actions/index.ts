export { Action } from "./action";
export * from "./user";
export * from "./drawing";
