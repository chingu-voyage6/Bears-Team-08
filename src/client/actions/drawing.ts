import { PaintKind, ID } from "@shared/contract";
import { Drawing } from "@shared/drawing";
import { Paint } from "@shared/paint";

import { Action, asErr, asReq, asRes, dispatcher, ThunkAction } from "./action";
import * as Api from "../api";

export type GetDrawingsReq = {
  token?: string;
  filters?: any[];
  offset: number;
  limit: number;
};
export type GetDrawingsRes = { drawings: Drawing[] };
export type GetDrawings = ThunkAction<
  "GET_DRAWINGS_REQUEST",
  "GET_DRAWINGS_SUCCESS",
  "GET_DRAWINGS_ERROR",
  GetDrawingsReq,
  GetDrawingsRes
>;

export const getDrawings = dispatcher(Api.getDrawings)<GetDrawings>(
  asReq("GET_DRAWINGS_REQUEST"),
  asRes("GET_DRAWINGS_SUCCESS"),
  asErr("GET_DRAWINGS_ERROR")
);

export type ChangePaintMethod = {
  type: "CHANGE_PAINT_METHOD";
  method: PaintKind;
};

export const changePaintMethod = (method: PaintKind): Action => ({
  type: "CHANGE_PAINT_METHOD",
  method
});

export type UndoReq = { token: string };
export type UndoRes = {};
export type Undo = ThunkAction<
  "UNDO_REQUEST",
  "UNDO_SUCCESS",
  "UNDO_ERROR",
  UndoReq,
  UndoRes
>;

export const undo = dispatcher(Api.undo)<Undo>(
  asReq("UNDO_REQUEST"),
  asRes("UNDO_SUCCESS"),
  asErr("UNDO_ERROR")
);

export type RedoReq = { token: string };
export type RedoRes = {};
export type Redo = ThunkAction<
  "REDO_REQUEST",
  "REDO_SUCCESS",
  "REDO_ERROR",
  UndoReq,
  UndoRes
>;

export const redo = dispatcher(Api.redo)<Redo>(
  asReq("REDO_REQUEST"),
  asRes("REDO_SUCCESS"),
  asErr("REDO_ERROR")
);

export type CreateDrawingReq = { name: string; token: string };
export type CreateDrawingRes = { drawing: Drawing };
export type CreateDrawing = ThunkAction<
  "CREATE_DRAWING_REQUEST",
  "CREATE_DRAWING_SUCCESS",
  "CREATE_DRAWING_ERROR",
  CreateDrawingReq,
  CreateDrawingRes
>;

export const createDrawing = dispatcher(Api.newDrawing)<CreateDrawing>(
  asReq("CREATE_DRAWING_REQUEST"),
  asRes("CREATE_DRAWING_SUCCESS"),
  asErr("CREATE_DRAWING_ERROR")
);

export type LoadDrawingReq = { id: ID; token: string };
export type LoadDrawingRes = { drawing: Drawing };
export type LoadDrawing = ThunkAction<
  "LOAD_DRAWING_REQUEST",
  "LOAD_DRAWING_SUCCESS",
  "LOAD_DRAWING_ERROR",
  LoadDrawingReq,
  LoadDrawingRes
>;

export const loadDrawing = dispatcher(Api.loadDrawing)<LoadDrawing>(
  asReq("LOAD_DRAWING_REQUEST"),
  asRes("LOAD_DRAWING_SUCCESS"),
  asErr("LOAD_DRAWING_ERROR")
);

export type SaveDrawingReq = { drawing: Drawing; token: string };
export type SaveDrawingRes = {};
export type SaveDrawing = ThunkAction<
  "SAVE_DRAWING_REQUEST",
  "SAVE_DRAWING_SUCCESS",
  "SAVE_DRAWING_ERROR",
  {},
  {}
>;

export const saveDrawing = dispatcher(Api.saveDrawing)<SaveDrawing>(
  asReq("SAVE_DRAWING_REQUEST"),
  asRes("SAVE_DRAWING_SUCCESS"),
  asErr("SAVE_DRAWING_ERROR")
);

export type AddPaintReq = { paint: Paint; token: string };
export type AddPaintRes = {};
export type AddPaint = ThunkAction<
  "ADD_PAINT_REQUEST",
  "ADD_PAINT_SUCCESS",
  "ADD_PAINT_ERROR",
  AddPaintReq,
  AddPaintRes
>;

export const addPaint = dispatcher(Api.addPaint)<AddPaint>(
  asReq("ADD_PAINT_REQUEST"),
  asRes("ADD_PAINT_SUCCESS"),
  asErr("ADD_PAINT_ERROR")
);

export type ModifyPaintReq = { token: string };
export type ModifyPaintRes = {};
export type ModifyPaint = ThunkAction<
  "MODIFY_PAINT_REQUEST",
  "MODIFY_PAINT_SUCCESS",
  "MODIFY_PAINT_ERROR",
  ModifyPaintReq,
  ModifyPaintRes
>;

export const modifyPaint = dispatcher(Api.modifyPaint)<ModifyPaint>(
  asReq("MODIFY_PAINT_REQUEST"),
  asRes("MODIFY_PAINT_SUCCESS"),
  asErr("MODIFY_PAINT_ERROR")
);

export type DeletePaintReq = { token: string };
export type DeletePaintRes = {};
export type DeletePaint = ThunkAction<
  "DELETE_PAINT_REQUEST",
  "DELETE_PAINT_SUCCESS",
  "DELETE_PAINT_ERROR",
  DeletePaintReq,
  DeletePaintRes
>;

export const deletePaint = dispatcher(Api.deletePaint)<DeletePaint>(
  asReq("DELETE_PAINT_REQUEST"),
  asRes("DELETE_PAINT_SUCCESS"),
  asErr("DELETE_PAINT_ERROR")
);
