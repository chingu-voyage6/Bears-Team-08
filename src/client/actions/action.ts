type Req<T> = { request: T };
type Res<T> = { response: T };
type Err = { error: string };

import * as Drawing from "./drawing";
import * as User from "./user";

export type ThunkAction<ReqKind, ResKind, ErrKind, _Req, _Res> =
  | ({ type: ReqKind } & Req<_Req>)
  | ({ type: ResKind } & Req<_Req> & Res<_Res>)
  | ({ type: ErrKind } & Req<_Req> & Err);

export type Action =
  | Drawing.AddPaint
  | Drawing.ChangePaintMethod
  | Drawing.DeletePaint
  | Drawing.GetDrawings
  | Drawing.LoadDrawing
  | Drawing.ModifyPaint
  | Drawing.CreateDrawing
  | Drawing.Redo
  | Drawing.SaveDrawing
  | Drawing.Undo
  | User.LoginUser
  | User.LogoutUser;

type ActionKind = Action["type"];

// TypeScript won't narrow a `ThunkAction` union directly, but
// we can help it out by tagging the three permutations.
export const asReq = <ReqKind extends ActionKind>(type: ReqKind) => <_Req>(
  request: _Req
) => ({
  type,
  request
});

export const asRes = <ResKind extends ActionKind>(type: ResKind) => <
  _Req,
  _Res
>(
  request: _Req,
  response: _Res
) => ({ type, request, response });

export const asErr = <ErrKind extends ActionKind>(type: ErrKind) => <_Req>(
  request: _Req,
  error: string
) => ({ type, request, error });

type Dispatch<A> = (a: A) => A;
type Resolve<_Req, _Res> = (request: _Req) => Promise<_Res>;

type ReqCreator<A, _Req> = (req: _Req) => A;
type ResCreator<A, _Req, _Res> = (req: _Req, res: _Res) => A;
type ErrCreator<A, _Req> = (req: _Req, err: string) => A;

export const dispatcher = <_Req, _Res>(fn: Resolve<_Req, _Res>) => <
  A extends Action
>(
  req: ReqCreator<A, _Req>,
  res: ResCreator<A, _Req, _Res>,
  err: ErrCreator<A, _Req>
) => (request: _Req) => (dispatch: Dispatch<A>) => {
  dispatch(req(request));
  fn(request)
    .then(response => dispatch(res(request, response)))
    .catch(e => dispatch(err(request, e.message)));
};
