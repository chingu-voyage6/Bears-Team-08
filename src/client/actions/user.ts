import { AuthUserJSON, LoginUserJSON } from "@shared/contract";

import { Action, asErr, asReq, asRes, dispatcher, ThunkAction } from "./action";
import * as Api from "../api";

export type LoginUserReq = LoginUserJSON;
export type LoginUserRes = AuthUserJSON;
export type LoginUser = ThunkAction<
  "LOGIN_USER_REQUEST",
  "LOGIN_USER_SUCCESS",
  "LOGIN_USER_ERROR",
  LoginUserReq,
  LoginUserRes
>;

export const loginUser = dispatcher(Api.loginUser)<LoginUser>(
  asReq("LOGIN_USER_REQUEST"),
  asRes("LOGIN_USER_SUCCESS"),
  asErr("LOGIN_USER_ERROR")
);

export type LogoutUser = {
  type: "LOGOUT_USER";
};

export const logoutUser = (): Action => ({
  type: "LOGOUT_USER"
});
