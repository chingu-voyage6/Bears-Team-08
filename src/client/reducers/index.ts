import { combineReducers, Reducer } from "redux";

import { Action } from "../actions";
import { PaintKind, UserJSON } from "@shared/contract";
import { Drawing } from "@shared/drawing";

const paintMethod = (state: PaintKind = PaintKind.Freehand, action: Action) => {
  switch (action.type) {
    case "CHANGE_PAINT_METHOD":
      return action.method;
    default:
      return state;
  }
};

const isDrawing = (state: boolean = false, action: Action): boolean => {
  switch (action.type) {
    case "CREATE_DRAWING_REQUEST":
      return true;
    default:
      return state;
  }
};

const drawing = (state: Drawing = null, action: Action): Drawing => {
  switch (action.type) {
    case "CREATE_DRAWING_REQUEST": {
      return new Drawing(action.request.name);
    }
    case "CREATE_DRAWING_SUCCESS": {
      return action.response.drawing;
    }
    default:
      return state;
  }
};

const drawings = (state: Drawing[] = [], action: Action): Drawing[] => {
  switch (action.type) {
    default:
      return state;
  }
};

const isSaving = (state: boolean = false, action: Action): boolean => {
  switch (action.type) {
    case "CREATE_DRAWING_REQUEST":
    case "SAVE_DRAWING_REQUEST":
      return true;
    case "CREATE_DRAWING_SUCCESS":
    case "CREATE_DRAWING_ERROR":
    case "SAVE_DRAWING_SUCCESS":
    case "SAVE_DRAWING_ERROR":
      return false;
    default:
      return state;
  }
};

const isLoading = (state: boolean = false, action: Action): boolean => {
  switch (action.type) {
    case "LOAD_DRAWING_REQUEST":
      return true;
    case "LOAD_DRAWING_SUCCESS":
    case "LOAD_DRAWING_ERROR":
      return false;
    default:
      return state;
  }
};

const error = (state: string = "", action: Action): string => {
  switch (action.type) {
    case "LOAD_DRAWING_ERROR":
    case "LOGIN_USER_ERROR":
    case "SAVE_DRAWING_ERROR":
      return action.error;
    default:
      return state;
  }
};

const token = (state: string = "", action: Action): string => {
  switch (action.type) {
    case "LOGIN_USER_SUCCESS":
      return action.response.token;
    default:
      return state;
  }
};

const user = (state: UserJSON = {}, action: Action): UserJSON => {
  switch (action.type) {
    case "LOGIN_USER_SUCCESS":
      return action.response.user;
    default:
      return state;
  }
};

export type State = {
  drawing?: Drawing;
  drawings: Drawing[];
  error: string;
  isDrawing: boolean;
  isLoading: boolean;
  isSaving: boolean;
  paintMethod: PaintKind;
  token?: string;
  user?: UserJSON;
};

export const initialState: State = {
  drawing: null,
  drawings: [],
  error: "",
  isDrawing: false,
  isLoading: false,
  isSaving: false,
  paintMethod: PaintKind.Freehand,
  token: "",
  user: {}
};

export const rootReducer = combineReducers<State>({
  drawing,
  drawings,
  error,
  isDrawing,
  isLoading,
  isSaving,
  paintMethod,
  token,
  user
});
