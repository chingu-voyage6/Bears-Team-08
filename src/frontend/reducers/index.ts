import { combineReducers, Reducer } from "redux";

import { Action } from "../actions";
import { PaintKind } from "../../shared/paint";
import { Drawing } from "../../shared/drawing";

const method = (state: PaintKind = PaintKind.Freehand, action: Action) => {
  switch (action.type) {
    case "CHANGE_PAINT_METHOD":
      return action.method;
    default:
      return state;
  }
};

const drawing = (state: Drawing = null, action: Action) => {
  switch (action.type) {
    case "NEW_DRAWING_REQUEST": {
      return new Drawing(action.request.name);
    }
    default:
      return state;
  }
};

const isSaving = (state: boolean = false, action: Action) => {
  switch (action.type) {
    case "NEW_DRAWING_REQUEST":
    case "SAVE_DRAWING_REQUEST":
      return true;
    case "NEW_DRAWING_SUCCESS":
    case "NEW_DRAWING_ERROR":
    case "SAVE_DRAWING_SUCCESS":
    case "SAVE_DRAWING_ERROR":
      return false;
    default:
      return state;
  }
};

const isLoading = (state: boolean = false, action: Action): boolean => {
  switch (action.type) {
    case "LOAD_DRAWING_REQUEST":
      return true;
    case "LOAD_DRAWING_SUCCESS":
    case "LOAD_DRAWING_ERROR":
      return false;
    default:
      return state;
  }
};

const error = (state: string = "", action: Action): string => {
  switch (action.type) {
    case "LOAD_DRAWING_REQUEST":
    case "SAVE_DRAWING_REQUEST":
      return "";
    case "LOAD_DRAWING_ERROR":
    case "SAVE_DRAWING_ERROR":
      return action.error;
    default:
      return state;
  }
};

export type State = {
  method: PaintKind;
  drawing?: Drawing;
  isSaving: boolean;
  isLoading: boolean;
  error: string;
};

export const initialState: State = {
  method: PaintKind.Freehand,
  drawing: null,
  isSaving: false,
  isLoading: false,
  error: ""
};

export const rootReducer = combineReducers<State>({
  method,
  drawing,
  isSaving,
  isLoading,
  error
});
