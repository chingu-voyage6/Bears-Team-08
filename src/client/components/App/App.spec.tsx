import * as React from "react";
import * as Renderer from "react-test-renderer";
import { mount, ReactWrapper } from "enzyme";
import { Store } from "redux";

import { App } from ".";
import { DrawingBoard, Explorer, Navbar, Sidebar } from "../";
import { wrapComponentInProvider, createStore } from "@shared/testUtil";
import { rootReducer, initialState } from "../../reducers";
import { newDrawing, changePaintMethod } from "../../actions";

describe("components/App", () => {
  it("renders", () => {
    const store = createStore();
    const tree = Renderer.create(wrapComponentInProvider(<App />, store));
    expect(tree).toMatchSnapshot();
  });

  describe("clicking on new drawing button", () => {
    const drawingName = "Hello World Drawing";
    let store: Store;
    let wrapper: ReactWrapper;

    beforeEach(() => {
      store = createStore();
      newDrawing({ name: drawingName })(store.dispatch);
      wrapper = mount(wrapComponentInProvider(<App />, store));
    });

    it("should update current active drawing name in navbar", () => {
      expect(wrapper.find(Navbar).text()).toContain(drawingName);
    });

    it("should render Sidebar", () => {
      expect(wrapper.find(Sidebar).length).toEqual(1);
      expect(wrapper.find(Explorer).length).toEqual(0);
    });

    it("should render DrawingBoard", () => {
      expect(wrapper.find(DrawingBoard).length).toEqual(1);
      expect(wrapper.find(Explorer).length).toEqual(0);
    });
  });
});
