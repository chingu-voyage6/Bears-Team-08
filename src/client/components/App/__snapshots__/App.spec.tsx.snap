// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/App renders 1`] = `
<div
  className="undefined undefined"
>
  <section />
  <div>
    <label>
      Name:
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <button
      onClick={[Function]}
    >
      New Drawing
    </button>
  </div>
</div>
`;
